src/chain.c

```diff
@@ -57,23 +58,73 @@ qsort_cmp(const void *a, const void *b) {
  * Chain
  */
 
+
+static void
+hsk_chain_log(const hsk_chain_t *chain, const char *fmt, ...) {
+  printf("chain (%u): ", (uint32_t)chain->height);
+
+  va_list args;
+  va_start(args, fmt);
+  vprintf(fmt, args);
+  va_end(args);
+}
+
 int
-hsk_chain_init(hsk_chain_t *chain, const hsk_timedata_t *td) {
-  if (!chain || !td)
+hsk_chain_init(hsk_chain_t *chain, const hsk_timedata_t *td, const uv_loop_t *loop) {
+  if (!chain || !td || !loop)
     return HSK_EBADARGS;
 
   chain->height = 0;
   chain->tip = NULL;
   chain->genesis = NULL;
   chain->synced = false;
   chain->td = (hsk_timedata_t *)td;
+  chain->loop = (uv_loop_t *)loop;
 
   hsk_map_init_hash_map(&chain->hashes, free);
   hsk_map_init_int_map(&chain->heights, NULL);
   hsk_map_init_hash_map(&chain->orphans, free);
   hsk_map_init_hash_map(&chain->prevs, NULL);
 
-  return hsk_chain_init_genesis(chain);
+  chain->store = hsk_store_alloc(chain->loop);
+
+  if (!chain->store)
+    return HSK_ENOMEM;
+
+  int rc = hsk_store_open(chain->store);
+  if (rc != HSK_SUCCESS) {
+    return rc;
+  }
+
+  rc = hsk_chain_init_genesis(chain);
+  if (rc != HSK_SUCCESS) {
+    return rc;
+  }
+
+  // TODO: hardcoded 65536 because we don't know how many headers we've written
+  // So we read until we hit a header which has garbage data and break
+  // Here we're checking if bits == 0
+  for (int i = 1; i < 65536; i++) {
+    hsk_header_t *hdr = hsk_header_alloc();
+
+    if (!hdr)
+      return HSK_ENOMEM;
+
+    hsk_store_read(chain->store, i, hdr);
+
+    // FIXME: ugly hack ugh!
+    if (hdr->bits == 0) {
```

Comment:
Does this come from `ftruncate` expanding file with `0`s ?

Reply:
Yep, this is to check if we're seeking above the size we've written. If we're reading a null byte block, because we're going above the written size, we will parse a header with bits = 0 (among other fields).

I think the one way to fix this hack is to create a file with size = number of headers * sizeof header, then we can os.Stat on it to get the exact number of headers we need to be seeking to.